hls-plugin-api/src/Ide/Plugin/Error.hs

```diff
@@ -0,0 +1,163 @@
+    -- takes a low precedence (3) in being returned as a response to the client.
+  | PluginDependencyFailed T.Text
+    -- |PluginRequestRefused allows your handler to inspect a request before
+    -- rejecting it. In effect it allows your plugin to act make a secondary
+    -- `pluginEnabled` decision after receiving the request. This should only be
+    -- used if the decision to accept the request can not be made in
+    -- `pluginEnabled`.
+    --
+    -- This error will be logged together with other errors of the same type
+    -- with Info. If it's the only response to a request, HLS will respond as if
+    -- no plugins passed the `pluginEnabled` stage.
+  | PluginRequestRefused
+    -- |PluginRuleFailed should be thrown when a Rule your response depends on
+    -- fails.
+    --
+    -- This error takes the name of the Rule that failed.
+    --
+    -- This error will be logged together with other errors of the same type
+    -- with Info, and will be converted into a ContentModified response code. It
+    -- takes a low precedence (3) in being returned as a response to the client.
+  | PluginRuleFailed T.Text
+    -- |PluginStaleResolve should be thrown when your resolve request is
+    -- provided with data it can no longer resolve.
+    --
+    -- This error will be logged individually with Info, and will be converted
+    -- into a ContentModified response. It takes a low precedence (3) in being
+    -- returned as a response to the client.
+  | PluginStaleResolve
+
+instance Pretty PluginError where
+    pretty = \case
+      PluginInternalError msg -> "Internal Error:" <+> pretty msg
+      PluginStaleResolve -> "Stale Resolve"
+      PluginRuleFailed rule       -> "Rule Failed:" <+> pretty rule
+      PluginInvalidParams text -> "Invalid Params:" <+> pretty text
+      PluginParseError text -> "Parse Error:" <+> pretty text
+      PluginInvalidRequest text -> "Invalid Request:" <+> pretty text
+      PluginDependencyFailed text -> "Dependency Failed:" <+> pretty text
+      PluginRequestRefused -> "Request Refused"
+
+handleMaybe :: Monad m => e -> Maybe b -> ExceptT e m b
+handleMaybe msg = maybe (throwE msg) return
+
+handleMaybeM :: Monad m => e -> m (Maybe b) -> ExceptT e m b
+handleMaybeM msg act = maybeM (throwE msg) return $ lift act
+
+hoistExceptT :: MonadIO m => ExceptT e IO a -> ExceptT e m a
+hoistExceptT = mapExceptT liftIO
+
+hoistMaybeT :: MonadIO m => MaybeT IO a -> MaybeT m a
+hoistMaybeT = mapMaybeT liftIO
+
+getNormalizedFilePathE :: Monad m => Uri -> ExceptT PluginError m NormalizedFilePath
+getNormalizedFilePathE uri = handleMaybe (PluginInvalidParams (T.pack $ "uriToNormalizedFile failed. Uri:" <>  show uri))
+        $ uriToNormalizedFilePath
+        $ toNormalizedUri uri
+
+-- ---------------------------------------------------------------------
+{- Note [PluginError]
+-- Each PluginError corresponds to either a specific ResponseError we want to
```

Comment:
The point of doing a multi-line comment is so you don't need the `--` at the start of each line :joy: